### PR TITLE
Implement TransactionalSession as part of the persistence

### PR DESCRIPTION
**Pull request by danielmarbach**

### CODE REVIEW

File: src/NServiceBus.Persistence.CosmosDB.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
+[assembly: System.Runtime.CompilerServices.InternalsVisibleTo(@"NServiceBus.Persistence.CosmosDB.TransactionalSession, PublicKey=0024000004800000940000000602000000240000525341310004000001000100dde965e6172e019ac82c2639ffe494dd2e7dd16347c34762a05732b492e110f2e4e2e1b5ef2d85c848ccfb671ee20a47c8d1376276708dc30a90ff1121b647ba3b7259a6bc383b2034938ef0e275b58b920375ac605076178123693c6c4f1331661a62eba28c249386855637780e3ff5f23a6d854700eaa6803ef48907513b92")]
+[assembly: System.Runtime.CompilerServices.InternalsVisibleTo(@"NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests, PublicKey=00240000048000009400000006020000002400005253413100040000010001007f16e21368ff041183fab592d9e8ed37e7be355e93323147a1d29983d6e591b04282e4da0c9e18bd901e112c0033925eb7d7872c2f1706655891c5c9d57297994f707d16ee9a8f40d978f064ee1ffc73c0db3f4712691b23bf596f75130f4ec978cf78757ec034625a5f27e6bb50c618931ea49f6f628fd74271c32959efb1c5")]
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:8]: what internals do the acceptance tests require?

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+using NServiceBus.AcceptanceTesting.Support;
+using NServiceBus.Configuration.AdvancedExtensibility;
+using NServiceBus.Persistence.CosmosDB;
+using NServiceBus.TransactionalSession.AcceptanceTests;
+
+public class ConfigureEndpointCosmosDBPersistence : IConfigureEndpointTestExecution
+{
+    public Task Configure(string endpointName, EndpointConfiguration configuration, RunSettings settings, PublisherMetadata publisherMetadata)
+    {
+        if (configuration.GetSettings().Get<bool>("Endpoint.SendOnly"))
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs:16]: why is this needed?

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests.csproj [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    <RootNamespace>NServiceBus.TransactionalSession.AcceptanceTests</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="GitHubActionsTestLogger" Version="2.0.1" />
+    <PackageReference Include="Microsoft.Azure.Cosmos" Version="3.30.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.1" />
+    <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
+    <PackageReference Include="Nunit" Version="3.13.3" />
+    <PackageReference Include="NServiceBus.AcceptanceTests.Sources" Version="8.0.0-rc.1" GeneratePathProperty="true" />
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests.csproj:18]: not a fan of bringin in the full sources package. I'd clearly prefer to only rely on the acceptance testing to avoid all the "compile remove" parts.

Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests.csproj:18]: We use this in this repo heavily. Also in other persistences. At this stage, this is almost an established practice

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs:16]: Aligned with the other Configure files in this repo

File: src/NServiceBus.Persistence.CosmosDB.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt (hunk 1 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:8]: Installers etc. that are in the Setup file

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs (hunk 2 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs:16]: Happy to remove though

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+/// <summary>
+/// Enables the transactional session feature.
+/// </summary>
+public static class CosmosTransactionalSessionExtensions
+{
+    /// <summary>
+    /// Enables transactional session for this endpoint.
+    /// </summary>
+    /// <param name="persistenceExtensions"></param>
+    /// <returns></returns>
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs:15]:
```suggestion
```

**Review by tmasternak — APPROVED** [review]

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        configuration.RegisterComponents(services => services.AddSingleton<IPartitionKeyFromHeadersExtractor, PartitionKeyProvider>());
+        configuration.RegisterComponents(services =>
+            services.AddSingleton<IContainerInformationFromHeadersExtractor, ContainerInformationProvider>());
+
+        return Task.CompletedTask;
+    }
+
+    public Task Cleanup() => Task.CompletedTask;
+
+    class PartitionKeyProvider : IPartitionKeyFromHeadersExtractor
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs:31]: should the tx session tests really use these default providers? Given the user has to specify the partition key on the options, this shouldn't be needed?

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    public class TransactionSessionDefaultServer : DefaultServer
+    {
+        public override Task<EndpointConfiguration> GetConfiguration(RunDescriptor runDescriptor, EndpointCustomizationConfiguration endpointConfiguration,
+            Func<EndpointConfiguration, Task> configurationBuilderCustomization) =>
+            base.GetConfiguration(runDescriptor, endpointConfiguration, async configuration =>
+            {
+                // Explicitly enforcing the type to be scanned. Otherwise the scanner would not pick it up in the acceptance tests
+                endpointConfiguration.TypesToInclude.Add(typeof(TransactionalSession));
+                endpointConfiguration.TypesToInclude.Add(typeof(CosmosTransactionalSession));
+
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs:18]: shouldn't this call `EnableTransactionalSession` somewhere (based on the template name)?

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    /// </summary>
+    public static class CosmosTransactionalSessionExtensions
+    {
+        /// <summary>
+        /// Enables transactional session for this endpoint.
+        /// </summary>
+        public static PersistenceExtensions<CosmosPersistence> EnableTransactionalSession(
+            this PersistenceExtensions<CosmosPersistence> persistenceExtensions)
+        {
+            persistenceExtensions.GetSettings().EnableFeatureByDefault<CosmosTransactionalSession>();
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs:17]: not that it matters too much, but is there are reason this is using `EnableFeatureByDefault` rather than `EnableFeature` given there is no valid use case to call both `EnableTransactionalSession()` AND `config.DisableFeature<CosmosTransactionalSession>()` ?

**Review by timbussmann — APPROVED** [review]

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs (hunk 5 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/ConfigureEndpointCosmosDBPersistence.cs:31]: On the incoming side it still requires a mapping for the actual messages that are delivered. All ATTs in this project are aligned to use the test run id as the partition key and a default container unless otherwise specified so I aligned that. The benefit is there is no mapping needed for the incoming messages in the actual tests

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs (hunk 7 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs:17]: Haven't thought about it from that angle. Good point. Will change to `EnableFeature`

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs (hunk 6 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs:18]: It is here https://github.com/Particular/NServiceBus.Persistence.CosmosDB/pull/417/files#diff-fff055b14a94240fcbf10059438c02faa41223f7183a90d3b372bd7b83e7e887R20 or maybe I misunderstood your comment

Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession.AcceptanceTests/TransactionSessionDefaultServer.cs:18]: Ah now I understand. You are saying it should be placed here instead of the Configure class. It would be doable if we introduce the concrete ConfigureXYZPersistence() methods that return the persistence extensions. I'm happy to go with this approach. Let me know if you prefer that

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs (hunk 7 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSessionExtensions.cs:17]: Ah EnableFeature is internal. So I would have to switch to `settings.Set(featureType.FullName, FeatureState.Enabled);`. Probably using the public API is the better choice here no?

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSession.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    sealed class CosmosTransactionalSession : Feature
+    {
+        public CosmosTransactionalSession()
+        {
+            Defaults(s =>
+            {
+                s.GetOrCreate<TransactionInformationConfiguration>().ExtractPartitionKeyFromHeaders(new ControlMessagePartitionKeyExtractor());
+                s.GetOrCreate<TransactionInformationConfiguration>().ExtractContainerInformationFromHeaders(new ControlMessageContainerInformationExtractor());
+
+                s.EnableFeatureByDefault<TransactionalSession>();
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSession.cs:16]: that call would be redundant now?

**Review by timbussmann — APPROVED** [review]
LGTM, one small comment

File: src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSession.cs (hunk 8 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB.TransactionalSession/CosmosTransactionalSession.cs]: Done
